In [1]:
# Clone the homework repository if it is not present
import os
import subprocess
repo_url = "https://github.com/mechristenson/aai-540-homework.git"
repo_path = "/home/ec2-user/SageMaker/MLOPSAssignments/aai-540-homework"
try:
    if os.path.isdir(os.path.join(repo_path, ".git")):
        print("Repository already exists. Skipping clone.")
    elif os.path.exists(repo_path):
        raise FileExistsError(
            f"Directory exists but is not a Git repository: {repo_path}"
        )
    else:
        subprocess.run(
            ["git", "clone", repo_url, repo_path],
            check=True
        )
        print("Repository cloned successfully.")
except Exception as e:
    print(f"Repository setup failed: {e}")
    raise

Repository already exists. Skipping clone.


In [2]:
# Inspect Homework 3.1 files
import os
homework_path = os.path.join(
    repo_path,
    "homework-3-1"
)
try:
    if not os.path.isdir(homework_path):
        raise FileNotFoundError(
            f"Homework 3.1 directory not found: {homework_path}"
        )
    print(f"Homework 3.1 directory found: {homework_path}")
    print("\nFiles and directories:")
    for item in sorted(os.listdir(homework_path)):
        item_path = os.path.join(homework_path, item)
        if os.path.isdir(item_path):
            print(f"[DIR]  {item}")
        else:
            print(f"[FILE] {item}")
except Exception as e:
    print(f"Homework 3.1 inspection failed: {e}")
    raise

Homework 3.1 directory found: /home/ec2-user/SageMaker/MLOPSAssignments/aai-540-homework/homework-3-1

Files and directories:
[FILE] housing.csv
[FILE] housing_gmaps_data_raw.csv


In [3]:
# Load the two datasets needed for the neighborhood feature group
import pandas as pd
housing_file = os.path.join(
    homework_path,
    "housing.csv"
)
gmaps_file = os.path.join(
    homework_path,
    "housing_gmaps_data_raw.csv"
)
try:
    # Check that both dataset files are available
    if not os.path.isfile(housing_file):
        raise FileNotFoundError(
            f"Housing dataset not found: {housing_file}"
        )
    if not os.path.isfile(gmaps_file):
        raise FileNotFoundError(
            f"Google Maps dataset not found: {gmaps_file}"
        )
    # Load the housing and Google Maps datasets
    housing_df = pd.read_csv(housing_file)
    gmaps_df = pd.read_csv(gmaps_file)
    print("Datasets loaded successfully.")
    # Show the basic information for the housing dataset
    print("\nHousing dataset:")
    print("Shape:", housing_df.shape)
    print("Columns:", list(housing_df.columns))
    # Show the basic information for the Google Maps dataset
    print("\nGoogle Maps dataset:")
    print("Shape:", gmaps_df.shape)
    print("Columns:", list(gmaps_df.columns))
except Exception as e:
    print(f"Dataset loading failed: {e}")
    raise

Datasets loaded successfully.

Housing dataset:
Shape: (20640, 10)
Columns: ['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'median_house_value', 'ocean_proximity']

Google Maps dataset:
Shape: (12590, 30)
Columns: ['street_number', 'route', 'locality-political', 'administrative_area_level_2-political', 'administrative_area_level_1-political', 'country-political', 'postal_code', 'address', 'longitude', 'latitude', 'neighborhood-political', 'postal_code_suffix', 'establishment-point_of_interest-transit_station', 'establishment-park-point_of_interest', 'premise', 'establishment-point_of_interest-subway_station-transit_station', 'airport-establishment-finance-moving_company-point_of_interest-storage', 'subpremise', 'bus_station-establishment-point_of_interest-transit_station', 'establishment-park-point_of_interest-tourist_attraction', 'establishment-natural_feature', 'airport-establishment-point_of_interest', 'p

In [4]:
# Inspect Key fields
# Check the fields we will be used for the neighborhood feature group
try:
    # Look at the housing fields needed for feature engineering
    print("Housing data sample:")
    display(
        housing_df[
            [
                "longitude",
                "latitude",
                "housing_median_age",
                "total_bedrooms",
                "households",
                "median_house_value",
                "ocean_proximity"
            ]
        ].head()
    )
    # Look at the Google Maps fields needed to connect addresses
    # with neighborhoods and postal codes
    print("\nGoogle Maps data sample:")
    display(
        gmaps_df[
            [
                "address",
                "postal_code",
                "neighborhood-political",
                "longitude",
                "latitude"
            ]
        ].head()
    )
    # Check missing values in the housing fields which will be used
    print("\nMissing values in housing data:")
    print(
        housing_df[
            [
                "housing_median_age",
                "total_bedrooms",
                "households",
                "median_house_value",
                "ocean_proximity"
            ]
        ].isnull().sum()
    )
    # Check missing values in the Google Maps fields which will be used
    print("\nMissing values in Google Maps data:")
    print(
        gmaps_df[
            [
                "postal_code",
                "neighborhood-political"
            ]
        ].isnull().sum()
    )
except Exception as e:
    print(f"Key field inspection failed: {e}")
    raise

Housing data sample:


,longitude,latitude,housing_median_age,total_bedrooms,households,median_house_value,ocean_proximity
0,-122.23,37.88,41.0,129.0,126.0,452600.0,NEAR BAY
1,-122.22,37.86,21.0,1106.0,1138.0,358500.0,NEAR BAY
2,-122.24,37.85,52.0,190.0,177.0,352100.0,NEAR BAY
3,-122.25,37.85,52.0,235.0,219.0,341300.0,NEAR BAY
4,-122.25,37.85,52.0,280.0,259.0,342200.0,NEAR BAY



Google Maps data sample:


,address,postal_code,neighborhood-political,longitude,latitude
0,"3130 Grizzly Peak Blvd, Berkeley, CA 94705, USA",94705.0,NaN,-122.23,37.88
1,"2005 Tunnel Rd, Oakland, CA 94611, USA",94611.0,Merriewood,-122.22,37.86
2,"6886 Chabot Rd, Oakland, CA 94618, USA",94618.0,Upper Rockridge,-122.24,37.85
3,"6365 Florio St, Oakland, CA 94618, USA",94618.0,Rockridge,-122.25,37.85
4,"5407 Bryant Ave, Oakland, CA 94618, USA",94618.0,Rockridge,-122.25,37.84



Missing values in housing data:
housing_median_age      0
total_bedrooms        207
households              0
median_house_value      0
ocean_proximity         0
dtype: int64

Missing values in Google Maps data:
postal_code                180
neighborhood-political    8413
dtype: int64


### Initial Data Inspection

The initial inspection confirms that both datasets contain the fields needed for the neighborhood feature group. The housing data contains the house level information such as median house value, median house age, households, total bedrooms, and ocean proximity. The Google Maps data provides the neighborhood and postal code information that we need to associate housing records with neighborhoods.

There are some missing values that need to be handled during feature engineering. The housing dataset has 207 missing values in `total_bedrooms`, while the Google Maps dataset has missing values in both `postal_code` and `neighborhood-political`.

Since the neighborhood feature group depends on correctly associating the housing records with the Google Maps neighborhood information, I will first check how well the two datasets can be matched using their longitude and latitude values.

In [5]:
# Check how the housing and Google Maps records match by location
try:
    # Keep only the location fields needed for the comparison
    housing_locations = housing_df[
        ["longitude", "latitude"]
    ].drop_duplicates()
    gmaps_locations = gmaps_df[
        ["longitude", "latitude"]
    ].drop_duplicates()
    # Check how many housing locations appear in the Google Maps data
    matched_locations = housing_locations.merge(
        gmaps_locations,
        on=["longitude", "latitude"],
        how="inner"
    )
    print("Housing records:", len(housing_df))
    print("Unique housing locations:", len(housing_locations))
    print("Unique Google Maps locations:", len(gmaps_locations))
    print(
        "Housing locations matched to Google Maps:",
        len(matched_locations)
    )
    # Check how many Google Maps records have a neighborhood
    # available for the matched locations
    gmaps_with_neighborhood = gmaps_df[
        gmaps_df["neighborhood-political"].notna()
    ]
    print(
        "\nGoogle Maps records with a neighborhood:",
        len(gmaps_with_neighborhood)
    )
    # Show a few matching records for verification
    print("\nSample matched locations:")
    display(
        housing_df[
            ["longitude", "latitude"]
        ].merge(
            gmaps_df[
                [
                    "longitude",
                    "latitude",
                    "postal_code",
                    "neighborhood-political"
                ]
            ],
            on=["longitude", "latitude"],
            how="inner"
        ).head(10)
    )
except Exception as e:
    print(f"Coordinate matching check failed: {e}")
    raise

Housing records: 20640
Unique housing locations: 12590
Unique Google Maps locations: 12590
Housing locations matched to Google Maps: 12590

Google Maps records with a neighborhood: 4177

Sample matched locations:


,longitude,latitude,postal_code,neighborhood-political
0,-122.23,37.88,94705.0,NaN
1,-122.22,37.86,94611.0,Merriewood
2,-122.24,37.85,94618.0,Upper Rockridge
3,-122.25,37.85,94618.0,Rockridge
4,-122.25,37.85,94618.0,Rockridge
5,-122.25,37.85,94618.0,Rockridge
6,-122.25,37.84,94618.0,Rockridge
7,-122.25,37.84,94618.0,Rockridge
8,-122.26,37.84,94618.0,Shafter
9,-122.25,37.84,94618.0,Rockridge


### Coordinate Matching Results

The coordinate comparison shows that the housing dataset contains 12,590 unique longitude and latitude combinations, and all 12,590 of these locations are also present in the Google Maps dataset. This confirms that longitude and latitude can be used to connect the housing records with the Google Maps information.

The Google Maps data contains neighborhood information for 4,177 records. Since `neighborhood-political` will be used as the primary key for the neighborhood feature group, records without a neighborhood cannot be used when creating the neighborhood-level features.

The next step is to combine the housing data with the Google Maps data using longitude and latitude. After the datasets are joined, I will keep the records that contain a neighborhood and postal code so the required neighborhood features can be created.

In [6]:
# Merge the datasets:
# Combine the housing data with the Google Maps neighborhood information
try:
    # Keep only the Google Maps fields needed for feature engineering
    gmaps_lookup = gmaps_df[
        [
            "longitude",
            "latitude",
            "postal_code",
            "neighborhood-political"
        ]
    ].drop_duplicates(
        subset=["longitude", "latitude"]
    )
    # Join the housing records with the Google Maps data
    housing_gmaps_df = housing_df.merge(
        gmaps_lookup,
        on=["longitude", "latitude"],
        how="left"
    )
    print("Datasets combined successfully.")
    print("Combined dataset shape:", housing_gmaps_df.shape)
    # Check how many housing records received neighborhood
    # and postal code information
    print("\nMissing location information:")
    print(
        housing_gmaps_df[
            [
                "postal_code",
                "neighborhood-political"
            ]
        ].isnull().sum()
    )
    # Keep records that have the neighborhood required for the feature group primary key
    neighborhood_df = housing_gmaps_df[
        housing_gmaps_df["neighborhood-political"].notna()
    ].copy()
    print(
        "\nRecords with neighborhood information:",
        len(neighborhood_df)
    )
    print(
        "Unique neighborhoods:",
        neighborhood_df["neighborhood-political"].nunique()
    )
    # Show a sample of the combined data
    print("\nSample combined records:")
    display(
        neighborhood_df[
            [
                "longitude",
                "latitude",
                "postal_code",
                "neighborhood-political",
                "housing_median_age",
                "total_bedrooms",
                "households",
                "median_house_value",
                "ocean_proximity"
            ]
        ].head(10)
    )
except Exception as e:
    print(f"Dataset merge failed: {e}")
    raise

Datasets combined successfully.
Combined dataset shape: (20640, 12)

Missing location information:
postal_code                 186
neighborhood-political    11640
dtype: int64

Records with neighborhood information: 9000
Unique neighborhoods: 1306

Sample combined records:


,longitude,latitude,postal_code,neighborhood-political,housing_median_age,total_bedrooms,households,median_house_value,ocean_proximity
1,-122.22,37.86,94611.0,Merriewood,21.0,1106.0,1138.0,358500.0,NEAR BAY
2,-122.24,37.85,94618.0,Upper Rockridge,52.0,190.0,177.0,352100.0,NEAR BAY
3,-122.25,37.85,94618.0,Rockridge,52.0,235.0,219.0,341300.0,NEAR BAY
4,-122.25,37.85,94618.0,Rockridge,52.0,280.0,259.0,342200.0,NEAR BAY
5,-122.25,37.85,94618.0,Rockridge,52.0,213.0,193.0,269700.0,NEAR BAY
6,-122.25,37.84,94618.0,Rockridge,52.0,489.0,514.0,299200.0,NEAR BAY
7,-122.25,37.84,94618.0,Rockridge,52.0,687.0,647.0,241400.0,NEAR BAY
8,-122.26,37.84,94618.0,Shafter,42.0,665.0,595.0,226700.0,NEAR BAY
9,-122.25,37.84,94618.0,Rockridge,52.0,707.0,714.0,261100.0,NEAR BAY
10,-122.26,37.85,94609.0,Upper Telegraph,52.0,434.0,402.0,281500.0,NEAR BAY


### Combined Housing and Neighborhood Data

After joining the datasets, the combined data contains all 20,640 original housing records and the additional postal code and neighborhood fields from the Google Maps data.

Neighborhood information is available for 9,000 housing records, representing 1,306 unique neighborhoods. These records will be used to create the neighborhood feature group because `neighborhood-political` is required as the primary key.

There are also some missing postal codes in the combined data. Before calculating bedrooms per household, I will check the missing values within the neighborhood records and determine whether the available postal code information can be used to impute missing bedroom values as required.

In [7]:
# check values needed for imputation
# Check missing values before creating the neighborhood features
try:
    # Check missing values in the fields used for the required features
    feature_fields = [
        "neighborhood-political",
        "postal_code",
        "housing_median_age",
        "total_bedrooms",
        "households",
        "median_house_value",
        "ocean_proximity"
    ]
    print("Missing values in neighborhood records:")
    print(
        neighborhood_df[
            feature_fields
        ].isnull().sum()
    )
    # Check the records where total bedrooms is missing
    missing_bedrooms = neighborhood_df[
        neighborhood_df["total_bedrooms"].isnull()
    ]
    print(
        "\nRecords with missing total bedrooms:",
        len(missing_bedrooms)
    )
    print(
        "Missing bedroom records with a postal code:",
        missing_bedrooms["postal_code"].notna().sum()
    )
    # Show a few records that will require bedroom imputation
    print("\nSample records with missing total bedrooms:")
    display(
        missing_bedrooms[
            [
                "neighborhood-political",
                "postal_code",
                "total_bedrooms",
                "households"
            ]
        ].head(10)
    )
except Exception as e:
    print(f"Missing value inspection failed: {e}")
    raise

Missing values in neighborhood records:
neighborhood-political     0
postal_code               11
housing_median_age         0
total_bedrooms            89
households                 0
median_house_value         0
ocean_proximity            0
dtype: int64

Records with missing total bedrooms: 89
Missing bedroom records with a postal code: 89

Sample records with missing total bedrooms:


,neighborhood-political,postal_code,total_bedrooms,households
290,Eastmont Hills,94605.0,NaN,218.0
341,Cox,94603.0,NaN,259.0
538,West Alameda,94501.0,NaN,1273.0
1606,Rossmoor,94595.0,NaN,626.0
2028,Roosevelt,93725.0,NaN,372.0
2115,Roosevelt,93727.0,NaN,1260.0
2578,Jacoby Creek,95521.0,NaN,669.0
2826,Hampton Woods,93312.0,NaN,16.0
3376,Tujunga,91042.0,NaN,769.0
3482,Sylmar,91342.0,NaN,623.0


### Missing Value Check

There are 89 neighborhood records with missing values for `total_bedrooms`. All 89 of these records have a postal code available, so the missing bedroom values can be imputed using the average `total_bedrooms` value for the corresponding postal code as required by the assignment.

There are also 11 records with a missing postal code, but none of the records with missing bedrooms are missing their postal code. This means the postal code based bedroom imputation can be completed without dropping any of the records that have missing bedroom values.

The next step is to calculate the average number of bedrooms for each postal code, fill the missing bedroom values, and then calculate bedrooms per household.

In [8]:
# Impute missing bedroom values using the postal code average
try:
    # Calculate the average total bedrooms for each postal code
    postal_code_bedroom_avg = neighborhood_df.groupby(
        "postal_code"
    )["total_bedrooms"].transform("mean")
    # Fill missing bedroom values with the average for the postal code
    neighborhood_df["total_bedrooms_imputed"] = (
        neighborhood_df["total_bedrooms"].fillna(
            postal_code_bedroom_avg
        )
    )
    # Calculate bedrooms per household using the imputed values
    neighborhood_df["bedrooms_per_household"] = (
        neighborhood_df["total_bedrooms_imputed"]
        / neighborhood_df["households"]
    )
    print("Bedroom imputation completed successfully.")
    print(
        "\nMissing total bedrooms before imputation:",
        neighborhood_df["total_bedrooms"].isnull().sum()
    )
    print(
        "Missing total bedrooms after imputation:",
        neighborhood_df[
            "total_bedrooms_imputed"
        ].isnull().sum()
    )
    print(
        "Missing bedrooms per household:",
        neighborhood_df[
            "bedrooms_per_household"
        ].isnull().sum()
    )
    # Show some of the records that originally had missing bedrooms
    print("\nSample imputed records:")
    display(
        neighborhood_df.loc[
            neighborhood_df["total_bedrooms"].isnull(),
            [
                "neighborhood-political",
                "postal_code",
                "total_bedrooms",
                "total_bedrooms_imputed",
                "households",
                "bedrooms_per_household"
            ]
        ].head(10)
    )
except Exception as e:
    print(f"Bedroom imputation failed: {e}")
    raise

Bedroom imputation completed successfully.

Missing total bedrooms before imputation: 89
Missing total bedrooms after imputation: 1
Missing bedrooms per household: 1

Sample imputed records:


,neighborhood-political,postal_code,total_bedrooms,total_bedrooms_imputed,households,bedrooms_per_household
290,Eastmont Hills,94605.0,NaN,331.270270,218.0,1.519588
341,Cox,94603.0,NaN,260.642857,259.0,1.006343
538,West Alameda,94501.0,NaN,470.700000,1273.0,0.369756
1606,Rossmoor,94595.0,NaN,618.000000,626.0,0.987220
2028,Roosevelt,93725.0,NaN,551.181818,372.0,1.481672
2115,Roosevelt,93727.0,NaN,614.739130,1260.0,0.487888
2578,Jacoby Creek,95521.0,NaN,577.333333,669.0,0.862980
2826,Hampton Woods,93312.0,NaN,449.166667,16.0,28.072917
3376,Tujunga,91042.0,NaN,460.761905,769.0,0.599170
3482,Sylmar,91342.0,NaN,666.647059,623.0,1.070059


### Bedroom Imputation Results

The postal code based imputation filled 88 of the 89 missing `total_bedrooms` values. One record still has a missing value after imputation, which indicates that its postal code does not have another available bedroom value that can be used to calculate the postal code average.

Before creating the neighborhood level features, I will inspect this remaining record and the other records associated with its postal code. This will help determine why the required postal code imputation could not be completed for that record.

In [9]:
# Inspect the bedroom value that could not be imputed
try:
    # Find records that are still missing bedrooms after imputation
    remaining_missing = neighborhood_df[
        neighborhood_df["total_bedrooms_imputed"].isnull()
    ]
    print(
        "Records still missing bedrooms after imputation:",
        len(remaining_missing)
    )
    print("\nRemaining record:")
    display(
        remaining_missing[
            [
                "neighborhood-political",
                "postal_code",
                "total_bedrooms",
                "total_bedrooms_imputed",
                "households"
            ]
        ]
    )
    # Check all neighborhood records with the same postal code
    if len(remaining_missing) > 0:
        missing_postal_codes = (
            remaining_missing["postal_code"]
            .dropna()
            .unique()
        )
        print("\nRecords with the same postal code:")
        display(
            neighborhood_df[
                neighborhood_df["postal_code"].isin(
                    missing_postal_codes
                )
            ][
                [
                    "neighborhood-political",
                    "postal_code",
                    "total_bedrooms",
                    "households"
                ]
            ]
        )
except Exception as e:
    print(f"Remaining missing value inspection failed: {e}")
    raise

Records still missing bedrooms after imputation: 1

Remaining record:


,neighborhood-political,postal_code,total_bedrooms,total_bedrooms_imputed,households
9845,Jack's Peak,93940.0,NaN,NaN,393.0



Records with the same postal code:


,neighborhood-political,postal_code,total_bedrooms,households
9845,Jack's Peak,93940.0,NaN,393.0


### Remaining Missing Bedroom Value

The remaining missing bedroom value belongs to Jack's Peak in postal code 93940. Within the neighborhood subset, this is the only record for postal code 93940, and its `total_bedrooms` value is missing. Therefore, a postal code average cannot be calculated from the neighborhood records alone.

Before using another imputation method, I will check the complete merged dataset for other housing records in postal code 93940. If valid bedroom values are available there, they can be used to calculate the required postal code average.

In [10]:
# Check the full merged dataset for the remaining postal code
try:
    remaining_postal_code = 93940.0
    postal_code_records = housing_gmaps_df[
        housing_gmaps_df["postal_code"] == remaining_postal_code
    ].copy()
    print(
        f"Records for postal code {int(remaining_postal_code)}:",
        len(postal_code_records)
    )
    print(
        "Records with available total bedrooms:",
        postal_code_records["total_bedrooms"].notna().sum()
    )
    print(
        "Records with missing total bedrooms:",
        postal_code_records["total_bedrooms"].isnull().sum()
    )
    print(
        "\nAverage total bedrooms for postal code:",
        postal_code_records["total_bedrooms"].mean()
    )
    print("\nRecords for this postal code:")
    display(
        postal_code_records[
            [
                "longitude",
                "latitude",
                "postal_code",
                "neighborhood-political",
                "total_bedrooms",
                "households"
            ]
        ]
    )
except Exception as e:
    print(f"Postal-code inspection failed: {e}")
    raise

Records for postal code 93940: 26
Records with available total bedrooms: 25
Records with missing total bedrooms: 1

Average total bedrooms for postal code: 514.32

Records for this postal code:


,longitude,latitude,postal_code,neighborhood-political,total_bedrooms,households
9789,-121.91,36.42,93940.0,NaN,261.0,171.0
9837,-121.90,36.61,93940.0,NaN,827.0,759.0
9838,-121.91,36.61,93940.0,NaN,597.0,554.0
9841,-121.90,36.60,93940.0,NaN,423.0,386.0
9842,-121.90,36.60,93940.0,NaN,649.0,601.0
9843,-121.90,36.60,93940.0,NaN,412.0,429.0
9844,-121.90,36.59,93940.0,NaN,510.0,459.0
9845,-121.90,36.58,93940.0,Jack's Peak,NaN,393.0
9846,-121.91,36.59,93940.0,NaN,335.0,322.0
9847,-121.91,36.60,93940.0,NaN,410.0,406.0


### Postal Code Imputation Check

The complete merged dataset contains 26 housing records for postal code 93940. Of these records, 25 have a valid `total_bedrooms` value, producing an average of 514.32 bedrooms for the postal code.

This confirms that the remaining missing value for Jack's Peak can be imputed using the postal code average as required. The earlier missing value occurred because the average was calculated only from records that already had neighborhood information. To avoid this problem, I will calculate postal code bedroom averages from the complete merged dataset and use those averages to fill missing bedroom values in the neighborhood records.

In [11]:
# Impute missing bedrooms using postal code averages from the full dataset
try:
    # Calculate average total bedrooms for each postal code
    # using all available records in the merged dataset
    postal_code_bedroom_avg = (
        housing_gmaps_df.groupby("postal_code")["total_bedrooms"]
        .mean()
    )
    # Keep the original bedroom values in a separate column
    neighborhood_df["total_bedrooms_imputed"] = (
        neighborhood_df["total_bedrooms"]
    )
    # Find the records that require imputation
    missing_bedrooms_mask = (
        neighborhood_df["total_bedrooms_imputed"].isnull()
    )
    # Fill missing bedrooms using the corresponding postal code average
    neighborhood_df.loc[
        missing_bedrooms_mask,
        "total_bedrooms_imputed"
    ] = (
        neighborhood_df.loc[
            missing_bedrooms_mask,
            "postal_code"
        ].map(postal_code_bedroom_avg)
    )
    # Calculate bedrooms per household
    neighborhood_df["bedrooms_per_household"] = (
        neighborhood_df["total_bedrooms_imputed"]
        / neighborhood_df["households"]
    )
    print("Final bedroom imputation completed successfully.")
    print(
        "\nRecords originally missing total bedrooms:",
        neighborhood_df["total_bedrooms"].isnull().sum()
    )
    print(
        "Missing bedrooms after imputation:",
        neighborhood_df[
            "total_bedrooms_imputed"
        ].isnull().sum()
    )
    print(
        "Missing bedrooms per household:",
        neighborhood_df[
            "bedrooms_per_household"
        ].isnull().sum()
    )
    # Verify the records that were imputed
    print("\nSample imputed records:")
    display(
        neighborhood_df.loc[
            neighborhood_df["total_bedrooms"].isnull(),
            [
                "neighborhood-political",
                "postal_code",
                "total_bedrooms",
                "total_bedrooms_imputed",
                "households",
                "bedrooms_per_household"
            ]
        ].head(10)
    )
    # Verify the Jack's Peak record specifically
    print("\nJack's Peak verification:")
    display(
        neighborhood_df[
            neighborhood_df["neighborhood-political"]
            == "Jack's Peak"
        ][
            [
                "neighborhood-political",
                "postal_code",
                "total_bedrooms",
                "total_bedrooms_imputed",
                "households",
                "bedrooms_per_household"
            ]
        ]
    )
except Exception as e:
    print(f"Final bedroom imputation failed: {e}")
    raise

Final bedroom imputation completed successfully.

Records originally missing total bedrooms: 89
Missing bedrooms after imputation: 0
Missing bedrooms per household: 0

Sample imputed records:


,neighborhood-political,postal_code,total_bedrooms,total_bedrooms_imputed,households,bedrooms_per_household
290,Eastmont Hills,94605.0,NaN,340.166667,218.0,1.560398
341,Cox,94603.0,NaN,274.968750,259.0,1.061655
538,West Alameda,94501.0,NaN,443.367347,1273.0,0.348285
1606,Rossmoor,94595.0,NaN,693.500000,626.0,1.107827
2028,Roosevelt,93725.0,NaN,459.111111,372.0,1.234170
2115,Roosevelt,93727.0,NaN,573.148148,1260.0,0.454879
2578,Jacoby Creek,95521.0,NaN,426.928571,669.0,0.638159
2826,Hampton Woods,93312.0,NaN,449.166667,16.0,28.072917
3376,Tujunga,91042.0,NaN,460.761905,769.0,0.599170
3482,Sylmar,91342.0,NaN,666.647059,623.0,1.070059



Jack's Peak verification:


,neighborhood-political,postal_code,total_bedrooms,total_bedrooms_imputed,households,bedrooms_per_household
9845,Jack's Peak,93940.0,NaN,514.32,393.0,1.308702


### Final Bedroom Imputation

The bedroom imputation is now complete. All 89 records that originally had missing `total_bedrooms` values were filled using the average bedroom value for their postal code. The postal code averages were calculated from the complete merged dataset so that all available housing records could contribute to the imputation.

After imputation, there are no missing values in `total_bedrooms_imputed` or `bedrooms_per_household`. For example, the missing bedroom value for Jack's Peak was filled with the postal code 93940 average of 514.32 bedrooms.

The data is now ready for the remaining neighborhood level feature engineering and aggregation.

In [12]:
# Create ocean proximity indicators
# Create one hot encoded ocean proximity features
try:
    # Check the ocean proximity categories before encoding
    print("Ocean proximity categories:")
    print(
        neighborhood_df["ocean_proximity"]
        .value_counts()
    )
    # Create the five required ocean proximity features
    neighborhood_df["less_than_1h_ocean"] = (
        neighborhood_df["ocean_proximity"] == "<1H OCEAN"
    ).astype(int)
    neighborhood_df["inland"] = (
        neighborhood_df["ocean_proximity"] == "INLAND"
    ).astype(int)
    neighborhood_df["island"] = (
        neighborhood_df["ocean_proximity"] == "ISLAND"
    ).astype(int)
    neighborhood_df["near_bay"] = (
        neighborhood_df["ocean_proximity"] == "NEAR BAY"
    ).astype(int)
    neighborhood_df["near_ocean"] = (
        neighborhood_df["ocean_proximity"] == "NEAR OCEAN"
    ).astype(int)
    print("\nOcean proximity features created successfully.")
    # Verify the one hot encoded features
    print("\nFeature totals:")
    print(
        neighborhood_df[
            [
                "less_than_1h_ocean",
                "inland",
                "island",
                "near_bay",
                "near_ocean"
            ]
        ].sum()
    )
    print("\nSample encoded records:")
    display(
        neighborhood_df[
            [
                "neighborhood-political",
                "ocean_proximity",
                "less_than_1h_ocean",
                "inland",
                "island",
                "near_bay",
                "near_ocean"
            ]
        ].head(10)
    )
except Exception as e:
    print(f"Ocean proximity encoding failed: {e}")
    raise

Ocean proximity categories:
ocean_proximity
<1H OCEAN     4651
INLAND        1494
NEAR OCEAN    1445
NEAR BAY      1410
Name: count, dtype: int64

Ocean proximity features created successfully.

Feature totals:
less_than_1h_ocean    4651
inland                1494
island                   0
near_bay              1410
near_ocean            1445
dtype: int64

Sample encoded records:


,neighborhood-political,ocean_proximity,less_than_1h_ocean,inland,island,near_bay,near_ocean
1,Merriewood,NEAR BAY,0,0,0,1,0
2,Upper Rockridge,NEAR BAY,0,0,0,1,0
3,Rockridge,NEAR BAY,0,0,0,1,0
4,Rockridge,NEAR BAY,0,0,0,1,0
5,Rockridge,NEAR BAY,0,0,0,1,0
6,Rockridge,NEAR BAY,0,0,0,1,0
7,Rockridge,NEAR BAY,0,0,0,1,0
8,Shafter,NEAR BAY,0,0,0,1,0
9,Rockridge,NEAR BAY,0,0,0,1,0
10,Upper Telegraph,NEAR BAY,0,0,0,1,0


### Ocean Proximity Features

The `ocean_proximity` field was converted into the five one hot encoded features required for the neighborhood feature group. Each record now indicates whether it is less than one hour from the ocean, inland, on an island, near the bay, or near the ocean.

The encoded feature totals account for all 9,000 neighborhood records. There are no island records in the neighborhood subset, so the `island` feature contains zeros for all records. The feature is still retained because it is part of the required feature group schema.

The next step is to aggregate the housing records by neighborhood and calculate the required neighborhood level values.

In [13]:
# Create one record for each neighborhood
import numpy as np
try:
    # Aggregate the required values by neighborhood
    neighborhood_features_df = (
        neighborhood_df.groupby(
            "neighborhood-political",
            as_index=False
        )
        .agg(
            less_than_1h_ocean=("less_than_1h_ocean", "max"),
            inland=("inland", "max"),
            island=("island", "max"),
            near_bay=("near_bay", "max"),
            near_ocean=("near_ocean", "max"),
            median_house_value=("median_house_value", "mean"),
            median_house_age=("housing_median_age", "mean"),
            total_households=("households", "mean"),
            bedrooms_per_household=("bedrooms_per_household", "mean")
        )
    )
    # Rename the neighborhood field for the Feature Store primary key
    neighborhood_features_df = neighborhood_features_df.rename(
        columns={
            "neighborhood-political": "neighborhood"
        }
    )
    # Cap the average median house value at 500,000
    neighborhood_features_df["median_house_value"] = (
        neighborhood_features_df["median_house_value"]
        .clip(upper=500000)
    )
    # Discretize average house age into groups of 10 years
    neighborhood_features_df["median_house_age"] = (
        np.floor(
            neighborhood_features_df["median_house_age"] / 10
        ) * 10
    ).astype(int)
    # Round the average number of households up to an integer
    neighborhood_features_df["total_households"] = (
        np.ceil(
            neighborhood_features_df["total_households"]
        )
    ).astype(int)
    print("Neighborhood features created successfully.")
    print(
        "\nNeighborhood feature group shape:",
        neighborhood_features_df.shape
    )
    print(
        "Unique neighborhoods:",
        neighborhood_features_df["neighborhood"].nunique()
    )
    print("\nSample neighborhood features:")
    display(
        neighborhood_features_df.head(10)
    )
    # Check the three neighborhoods required by the assignment
    required_neighborhoods = [
        "Brooktree",
        "Fisherman's Wharf",
        "Los Osos"
    ]
    print("\nRequired neighborhood records:")
    display(
        neighborhood_features_df[
            neighborhood_features_df["neighborhood"].isin(
                required_neighborhoods
            )
        ]
    )
except Exception as e:
    print(f"Neighborhood feature creation failed: {e}")
    raise

Neighborhood features created successfully.

Neighborhood feature group shape: (1306, 10)
Unique neighborhoods: 1306

Sample neighborhood features:


,neighborhood,less_than_1h_ocean,inland,island,near_bay,near_ocean,median_house_value,median_house_age,total_households,bedrooms_per_household
0,28 Palms,1,0,0,0,0,222200.000000,20,923,1.017335
1,Acorn Industrial,0,0,0,1,0,81300.000000,50,147,1.659864
2,Adams Hill,1,0,0,0,0,250733.333333,30,494,1.034649
3,Agua Mansa Industrial Corridor,0,1,0,0,0,112300.000000,10,516,1.102713
4,Al Tahoe,0,1,0,0,0,109180.000000,20,249,1.641739
5,Alamitos Beach,0,0,0,0,1,188194.117647,30,664,1.117409
6,Alessandro,0,1,0,0,0,68075.000000,30,409,1.055551
7,Alessandro Heights,0,1,0,0,0,315200.000000,0,1107,1.066847
8,Alhambra Triangle,0,1,0,0,0,88850.000000,40,382,1.114779
9,Alice,0,0,0,1,0,404300.000000,30,483,1.020704



Required neighborhood records:


,neighborhood,less_than_1h_ocean,inland,island,near_bay,near_ocean,median_house_value,median_house_age,total_households,bedrooms_per_household
130,Brooktree,1,0,0,0,0,257400.0,0,1438,0.336642
390,Fisherman's Wharf,0,0,0,1,0,500000.0,50,250,1.268000
604,Los Osos,0,0,0,0,1,221612.5,10,612,1.047885


### Neighborhood Feature Engineering Results

The housing records were aggregated into 1,306 unique neighborhoods. Each neighborhood now has the five required ocean proximity indicators along with the derived housing features.

The average median house value was capped at $500,000, the average house age was grouped into 10 year intervals, and the average number of households was rounded up to an integer. Bedrooms per household was calculated using the bedroom values after postal code based imputation.

The three neighborhoods required for the final Feature Store queries, Brooktree, Fisherman's Wharf, and Los Osos, are all present in the processed data. The next step is to prepare these neighborhood features for ingestion into SageMaker Feature Store.

In [14]:
# Add the event time required by SageMaker Feature Store
import time
import boto3
try:
    # Feature group name used for checking existing records
    feature_group_name = "neighborhood-feature-group"
    # Create the Feature Store runtime client
    featurestore_runtime_check = boto3.client(
        "sagemaker-featurestore-runtime"
    )
    # Check whether the feature group already contains data
    try:
        existing_response = featurestore_runtime_check.get_record(
            FeatureGroupName=feature_group_name,
            RecordIdentifierValueAsString="Brooktree"
        )
        existing_record = existing_response.get("Record", [])
        if existing_record:
            # Reuse the event time from the existing Feature Store data
            event_time_feature = next(
                feature
                for feature in existing_record
                if feature["FeatureName"] == "event_time"
            )
            ingestion_time = float(
                event_time_feature["ValueAsString"]
            )
            print(
                "Existing Feature Store data found."
            )
            print(
                "Reusing event time:",
                ingestion_time
            )
        else:
            ingestion_time = float(time.time())
            print(
                "No existing Feature Store data found."
            )
            print(
                "New event time created:",
                ingestion_time
            )
    except Exception:
        # The feature group may not exist during the first run
        ingestion_time = float(time.time())
        print(
            "Feature Store data not available yet."
        )
        print(
            "New event time created:",
            ingestion_time
        )
    # Add the event time to the neighborhood features
    neighborhood_features_df["event_time"] = ingestion_time
    # Arrange the columns in the order used for the feature group
    feature_store_df = neighborhood_features_df[
        [
            "neighborhood",
            "event_time",
            "less_than_1h_ocean",
            "inland",
            "island",
            "near_bay",
            "near_ocean",
            "median_house_value",
            "median_house_age",
            "total_households",
            "bedrooms_per_household"
        ]
    ].copy()
    print("\nFeature Store data prepared successfully.")
    print(
        "\nFeature Store shape:",
        feature_store_df.shape
    )
    print("\nColumn data types:")
    print(feature_store_df.dtypes)
    print("\nMissing values:")
    print(feature_store_df.isnull().sum())
    print("\nDuplicate neighborhood keys:")
    print(
        feature_store_df["neighborhood"]
        .duplicated()
        .sum()
    )
    print("\nSample Feature Store records:")
    display(feature_store_df.head(10))
    # Verify the three neighborhoods 
    required_neighborhoods = [
        "Brooktree",
        "Fisherman's Wharf",
        "Los Osos"
    ]
    print("\nRequired neighborhood records:")
    display(
        feature_store_df[
            feature_store_df["neighborhood"].isin(
                required_neighborhoods
            )
        ]
    )
except Exception as e:
    print(f"Feature Store data preparation failed: {e}")
    raise

Existing Feature Store data found.
Reusing event time: 1789846800.8865445

Feature Store data prepared successfully.

Feature Store shape: (1306, 11)

Column data types:
neighborhood               object
event_time                float64
less_than_1h_ocean          int64
inland                      int64
island                      int64
near_bay                    int64
near_ocean                  int64
median_house_value        float64
median_house_age            int64
total_households            int64
bedrooms_per_household    float64
dtype: object

Missing values:
neighborhood              0
event_time                0
less_than_1h_ocean        0
inland                    0
island                    0
near_bay                  0
near_ocean                0
median_house_value        0
median_house_age          0
total_households          0
bedrooms_per_household    0
dtype: int64

Duplicate neighborhood keys:
0

Sample Feature Store records:


,neighborhood,event_time,less_than_1h_ocean,inland,island,near_bay,near_ocean,median_house_value,median_house_age,total_households,bedrooms_per_household
0,28 Palms,1.789847e+09,1,0,0,0,0,222200.000000,20,923,1.017335
1,Acorn Industrial,1.789847e+09,0,0,0,1,0,81300.000000,50,147,1.659864
2,Adams Hill,1.789847e+09,1,0,0,0,0,250733.333333,30,494,1.034649
3,Agua Mansa Industrial Corridor,1.789847e+09,0,1,0,0,0,112300.000000,10,516,1.102713
4,Al Tahoe,1.789847e+09,0,1,0,0,0,109180.000000,20,249,1.641739
5,Alamitos Beach,1.789847e+09,0,0,0,0,1,188194.117647,30,664,1.117409
6,Alessandro,1.789847e+09,0,1,0,0,0,68075.000000,30,409,1.055551
7,Alessandro Heights,1.789847e+09,0,1,0,0,0,315200.000000,0,1107,1.066847
8,Alhambra Triangle,1.789847e+09,0,1,0,0,0,88850.000000,40,382,1.114779
9,Alice,1.789847e+09,0,0,0,1,0,404300.000000,30,483,1.020704



Required neighborhood records:


,neighborhood,event_time,less_than_1h_ocean,inland,island,near_bay,near_ocean,median_house_value,median_house_age,total_households,bedrooms_per_household
130,Brooktree,1.789847e+09,1,0,0,0,0,257400.0,0,1438,0.336642
390,Fisherman's Wharf,1.789847e+09,0,0,0,1,0,500000.0,50,250,1.268000
604,Los Osos,1.789847e+09,0,0,0,0,1,221612.5,10,612,1.047885


### Feature Store Data Validation

The final neighborhood dataset contains 1,306 records and 11 features, including the neighborhood primary key and the ingestion event time. The dataset contains no missing values and no duplicate neighborhood keys, so each neighborhood can be stored as a unique record in SageMaker Feature Store.

The three neighborhoods required for the assignment are also present in the final dataset. The event time was generated in Python and will be used as the Feature Store event time field during ingestion.

The processed data is now ready to be used to create the SageMaker Feature Group.

In [15]:
# AWS and SageMaker configuration
import boto3
import sagemaker
try:
    # Create the SageMaker session
    sagemaker_session = sagemaker.Session()
    # Get the AWS region and execution role
    region = boto3.Session().region_name
    role = sagemaker.get_execution_role()
    # Create AWS clients used for Feature Store
    sagemaker_client = boto3.client(
        "sagemaker",
        region_name=region
    )
    featurestore_runtime = boto3.client(
        "sagemaker-featurestore-runtime",
        region_name=region
    )
    # Use the SageMaker default bucket for the offline store
    bucket = sagemaker_session.default_bucket()
    print("AWS configuration completed successfully.")
    print("Region:", region)
    print("S3 bucket:", bucket)
    print("Execution role:", role)
except Exception as e:
    print(f"AWS configuration failed: {e}")
    raise

sagemaker.config INFO - Not applying SDK defaults from location: /etc/xdg/sagemaker/config.yaml
sagemaker.config INFO - Not applying SDK defaults from location: /home/ec2-user/.config/sagemaker/config.yaml


/home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages/sagemaker/__init__.py:86: SageMakerV2DeprecationWarning: You are using the SageMaker Python SDK v2, which is on path of deprecation. v3 is the actively developed major version.
See https://github.com/aws/sagemaker-python-sdk/blob/master/migration.md for the migration guide. Set SAGEMAKER_SUPPRESS_V2_WARNING=1 to silence this warning.
  warn_v2_deprecation()
You are using the SageMaker Python SDK v2, which is on path of deprecation. v3 is the actively developed major version.
See https://github.com/aws/sagemaker-python-sdk/blob/master/migration.md for the migration guide. Set SAGEMAKER_SUPPRESS_V2_WARNING=1 to silence this warning.


AWS configuration completed successfully.
Region: us-east-1
S3 bucket: sagemaker-us-east-1-614554462636
Execution role: arn:aws:iam::614554462636:role/LabRole


### AWS Configuration

The SageMaker session was initialized successfully in the AWS Learner Lab environment. The notebook is running in the `us-east-1` region and the SageMaker default S3 bucket and execution role were identified successfully.

The next step is to define the neighborhood feature group using the processed dataset.

In [16]:
# Define the neighborhood Feature Store feature group
from sagemaker.feature_store.feature_group import FeatureGroup
try:
    # Name of the Feature Store feature group
    feature_group_name = "neighborhood-feature-group"
    # Create the FeatureGroup object
    neighborhood_feature_group = FeatureGroup(
        name=feature_group_name,
        sagemaker_session=sagemaker_session
    )
    # Load the feature definitions from the final dataframe
    neighborhood_feature_group.load_feature_definitions(
        data_frame=feature_store_df
    )
    print("Feature group definition created successfully.")
    print("Feature group name:", feature_group_name)
    print("\nFeature definitions:")
    for feature in neighborhood_feature_group.feature_definitions:
        print(
            feature.feature_name,
            "-",
            feature.feature_type
        )
except Exception as e:
    print(f"Feature group definition failed: {e}")
    raise

Feature group definition created successfully.
Feature group name: neighborhood-feature-group

Feature definitions:
neighborhood - FeatureTypeEnum.STRING
event_time - FeatureTypeEnum.FRACTIONAL
less_than_1h_ocean - FeatureTypeEnum.INTEGRAL
inland - FeatureTypeEnum.INTEGRAL
island - FeatureTypeEnum.INTEGRAL
near_bay - FeatureTypeEnum.INTEGRAL
near_ocean - FeatureTypeEnum.INTEGRAL
median_house_value - FeatureTypeEnum.FRACTIONAL
median_house_age - FeatureTypeEnum.INTEGRAL
total_households - FeatureTypeEnum.INTEGRAL
bedrooms_per_household - FeatureTypeEnum.FRACTIONAL


### Feature Group Definition

The neighborhood feature group schema was created successfully from the processed dataframe. SageMaker identified the neighborhood primary key as a string, the event time and integer based features as integral values, and the continuous housing features as fractional values.

The inferred feature types match the data prepared for the neighborhood feature group. The next step is to create the feature group in SageMaker Feature Store using `neighborhood` as the record identifier and `event_time` as the event time feature.

In [17]:
# Create the neighborhood feature group if it does not already exist
from botocore.exceptions import ClientError
try:
    # S3 location used by the Feature Store offline store
    offline_store_uri = (
        f"s3://{bucket}/feature-store/neighborhood-feature-group"
    )
    # Check whether the feature group already exists
    try:
        existing_feature_group = (
            sagemaker_client.describe_feature_group(
                FeatureGroupName=feature_group_name
            )
        )
        print("Feature group already exists!!!!")
        print("Feature group:", feature_group_name)
        print(
            "Current status:",
            existing_feature_group["FeatureGroupStatus"]
        )
    except ClientError as e:
        error_code = e.response["Error"]["Code"]
        # Create the feature group only when it does not exist
        if error_code == "ResourceNotFound":
            print("Creating feature group...")
            print("Feature group:", feature_group_name)
            print("Offline store:", offline_store_uri)
            neighborhood_feature_group.create(
                s3_uri=offline_store_uri,
                record_identifier_name="neighborhood",
                event_time_feature_name="event_time",
                role_arn=role,
                enable_online_store=True
            )
            print(
                "\nFeature group creation request "
                "submitted successfully."
            )
        else:
            raise
except Exception as e:
    print(f"Feature group setup failed: {e}")
    raise

Feature group already exists!!!!
Feature group: neighborhood-feature-group
Current status: Created


### Feature Group Creation

The neighborhood feature group creation request was submitted successfully to SageMaker Feature Store. The feature group uses `neighborhood` as the record identifier and `event_time` as the event time feature. The online store is enabled for retrieving individual neighborhood records, and the offline store is configured in the SageMaker S3 bucket.

Feature group creation is asynchronous, so I will verify that the feature group reaches the `Created` status before ingesting the neighborhood records.

In [18]:
# Check the status of the neighborhood feature group
import time
try:
    while True:
        feature_group_description = (
            sagemaker_client.describe_feature_group(
                FeatureGroupName=feature_group_name
            )
        )
        feature_group_status = (
            feature_group_description["FeatureGroupStatus"]
        )
        print(
            "Feature group status:",
            feature_group_status
        )
        if feature_group_status == "Created":
            print(
                "\nFeature group is ready for data ingestion."
            )
            break
        if feature_group_status in [
            "CreateFailed",
            "DeleteFailed"
        ]:
            raise RuntimeError(
                f"Feature group entered status: "
                f"{feature_group_status}"
            )
        # Wait before checking the status again
        time.sleep(5)
except Exception as e:
    print(f"Feature group status check failed: {e}")
    raise

Feature group status: Created

Feature group is ready for data ingestion.


In [19]:
# Ingest the neighborhood records if the feature group is empty
try:
    # Use Brooktree to check whether the feature group
    # has already been populated
    check_response = featurestore_runtime.get_record(
        FeatureGroupName=feature_group_name,
        RecordIdentifierValueAsString="Brooktree"
    )
    existing_record = check_response.get("Record", [])
    if existing_record:
        print("Feature Store data already exists.")
        print("Skipping ingestion to avoid duplicate record versions.")
    else:
        print(
            "Records to ingest:",
            len(feature_store_df)
        )
        # Ingest the processed neighborhood dataframe
        neighborhood_feature_group.ingest(
            data_frame=feature_store_df,
            max_workers=3,
            wait=True
        )
        print(
            "\nNeighborhood feature data ingested successfully."
        )
except Exception as e:
    print(f"Feature data ingestion check failed: {e}")
    raise

Feature Store data already exists.
Skipping ingestion to avoid duplicate record versions.


### Feature Store Data Ingestion

The notebook checks the online Feature Store before ingestion to determine whether the neighborhood data has already been loaded. If the data is not present, all 1,306 neighborhood records are ingested into the feature group. If the data already exists, ingestion is skipped to avoid creating unnecessary duplicate record versions.

The next step is to verify the stored data by retrieving neighborhood records from the Feature Store online store.

In [20]:
# Verify a neighborhood record from the Feature Store online store
try:
    test_neighborhood = "Brooktree"
    response = featurestore_runtime.get_record(
        FeatureGroupName=feature_group_name,
        RecordIdentifierValueAsString=test_neighborhood
    )
    record = response.get("Record", [])
    if not record:
        raise ValueError(
            f"No Feature Store record found for {test_neighborhood}"
        )
    print(
        f"Feature Store record retrieved successfully for "
        f"{test_neighborhood}."
    )
    # Convert the Feature Store response into a readable dictionary
    record_values = {
        feature["FeatureName"]: feature["ValueAsString"]
        for feature in record
    }
    brooktree_record_df = pd.DataFrame(
        [record_values]
    )
    print("\nStored feature values:")
    display(brooktree_record_df)
except Exception as e:
    print(f"Feature Store record retrieval failed: {e}")
    raise

Feature Store record retrieved successfully for Brooktree.

Stored feature values:


,neighborhood,event_time,less_than_1h_ocean,inland,island,near_bay,near_ocean,median_house_value,median_house_age,total_households,bedrooms_per_household
0,Brooktree,1789846800.8865445,1,0,0,0,0,257400.0,0,1438,0.33664180048046527


### Online Feature Store Verification

The Brooktree record was retrieved successfully from the SageMaker Feature Store online store. The returned values match the neighborhood features calculated before ingestion, confirming that the feature group contains the expected data.


In [21]:
# Query the three neighborhoods 
try:
    required_neighborhoods = [
        "Brooktree",
        "Fisherman's Wharf",
        "Los Osos"
    ]
    queried_records = []
    for neighborhood in required_neighborhoods:
        response = featurestore_runtime.get_record(
            FeatureGroupName=feature_group_name,
            RecordIdentifierValueAsString=neighborhood
        )
        record = response.get("Record", [])
        if not record:
            print(
                f"No Feature Store record found for {neighborhood}."
            )
            continue
        # Convert the Feature Store response into a dictionary
        record_values = {
            feature["FeatureName"]: feature["ValueAsString"]
            for feature in record
        }
        queried_records.append(record_values)
        print(
            f"Feature Store record retrieved for {neighborhood}."
        )
    # Display all required neighborhood records together
    required_records_df = pd.DataFrame(
        queried_records
    )
    print("\nRequired Feature Store query results:")
    display(required_records_df)
except Exception as e:
    print(f"Required neighborhood query failed: {e}")
    raise

Feature Store record retrieved for Brooktree.
Feature Store record retrieved for Fisherman's Wharf.
Feature Store record retrieved for Los Osos.

Required Feature Store query results:


,neighborhood,event_time,less_than_1h_ocean,inland,island,near_bay,near_ocean,median_house_value,median_house_age,total_households,bedrooms_per_household
0,Brooktree,1789846800.8865445,1,0,0,0,0,257400.0,0,1438,0.33664180048046527
1,Fisherman's Wharf,1789846800.8865445,0,0,0,1,0,500000.0,50,250,1.268
2,Los Osos,1789846800.8865445,0,0,0,0,1,221612.5,10,612,1.0478845404823531


In [22]:
# Submit Week 3 assignment to GitHub
import os
import shutil
import subprocess
import getpass
def submit_week_to_github(week_name):
    repo = "/home/ec2-user/SageMaker/MLOPSAssignments-git"
    source_notebook = "/home/ec2-user/SageMaker/Untitled.ipynb"
    week_dir = os.path.join(
        repo,
        week_name
    )
    destination_notebook = os.path.join(
        week_dir,
        f"{week_name}.ipynb"
    )
    try:
        # Validate the Git repository
        if not os.path.isdir(
            os.path.join(repo, ".git")
        ):
            raise FileNotFoundError(
                f"Git repository not found: {repo}"
            )
        # Validate the current notebook
        if not os.path.isfile(source_notebook):
            raise FileNotFoundError(
                f"Notebook not found: {source_notebook}"
            )
        # Create the Week 3 directory if needed
        os.makedirs(
            week_dir,
            exist_ok=True
        )
        # Copy the current notebook into the Week 3 directory
        shutil.copy2(
            source_notebook,
            destination_notebook
        )
        print(
            f"Notebook copied to "
            f"{week_name}/{week_name}.ipynb"
        )
        # Check Git status
        status = subprocess.run(
            ["git", "status", "--short"],
            cwd=repo,
            capture_output=True,
            text=True,
            check=True
        )
        print("\nGit status:")
        print(
            status.stdout
            if status.stdout
            else "No changes detected."
        )
        # Stage the Week 3 directory
        subprocess.run(
            ["git", "add", week_name],
            cwd=repo,
            check=True
        )
        # Commit the Week 3 assignment
        commit = subprocess.run(
            [
                "git",
                "commit",
                "-m",
                f"Add {week_name} assignment"
            ],
            cwd=repo,
            capture_output=True,
            text=True
        )
        if commit.returncode == 0:
            print("\nCommit created successfully.")
            print(commit.stdout)
        elif "nothing to commit" in (
            commit.stdout + commit.stderr
        ).lower():
            print("\nNothing new to commit.")
        else:
            raise RuntimeError(
                commit.stderr or commit.stdout
            )
        # Get GitHub credentials securely
        username = "sanjay0288"
        token = getpass.getpass(
            "Enter GitHub Personal Access Token: "
        )
        remote_url = (
            f"https://{username}:{token}"
            f"@github.com/{username}/"
            f"MLOPSAssignments.git"
        )
        # Push the assignment to GitHub
        subprocess.run(
            [
                "git",
                "push",
                remote_url,
                "main"
            ],
            cwd=repo,
            check=True
        )
        print(
            f"\n{week_name} assignment pushed "
            f"to GitHub successfully."
        )
    except Exception as e:
        print(f"GitHub submission failed: {e}")
        raise
submit_week_to_github("week3")

Notebook copied to week3/week3.ipynb

Git status:
 M week3/week3.ipynb


Commit created successfully.
[main 84589ca] Add week3 assignment
 Committer: EC2 Default User <ec2-user@ip-172-21-224-236.ec2.internal>
Your name and email address were configured automatically based
on your username and hostname. Please check that they are accurate.
You can suppress this message by setting them explicitly:

    git config --global user.name "Your Name"
    git config --global user.email you@example.com

After doing this, you may fix the identity used for this commit with:

    git commit --amend --reset-author

 1 file changed, 112 insertions(+), 2184 deletions(-)



Enter GitHub Personal Access Token:  ········



week3 assignment pushed to GitHub successfully.


To https://github.com/sanjay0288/MLOPSAssignments.git
   840a554..84589ca  main -> main
